ISPN-9893 New API indexing annotations

fax4ever opened the pull request:
https://github.com/tristantarrant/infinispan/pull/3
https://github.com/tristantarrant/infinispan/pull/4

to the current Infinispan main.

pruivo commented on api/pom.xml:
```diff
@@ -17,6 +17,11 @@
       </dependency>
+      <dependency>
+         <groupId>org.hibernate.search</groupId>
+         <artifactId>hibernate-search-mapper-pojo-base</artifactId>
```
wasn't supposed the new annotation to be independent of HSearch annotations?

fax4ever replied:
Great question! Theoretically you shouldn't have any Hibernate Search 6 dependencies on our own annotations, I totally agree with that.
The real problem here is that at the moment the only way to get the annotations processed by Hibernate Search is to annotate them with `@PropertyMapping` or `@TypeMapping` or ...
I had a chat with @yrodiere a while ago and without a significant investment on Search that's the only way to do that.
In my opinion we should accept this compromise and eventually we can remove it when Search is ready for it.
In the meantime I'm adding a note on top of this dependency.

yrodiere replied:
FWIW, the discussion: https://hibernate.zulipchat.com/#narrow/stream/132092-hibernate-search-dev/topic/custom.20annotations

fax4ever commented:
Thank @pruivo for the review. I did some stuff:

* Rebased on top of master
* Added a comment on top of the Hibernate Search 6 dependency
* Changed the `FullText` annotation to `Text`, I had a chat with @karesti yesterday, probably less is more...

pruivo commented on api/src/main/java/org/infinispan/api/annotations/indexing/model/Point.java:
```diff
@@ -0,0 +1,35 @@
+ * <p>
+ * Simplified version for Infinispan of {@link org.hibernate.search.engine.spatial.GeoPoint}
+ */
+public class Point {
```
make class and fields `final`?

fax4ever commented:
Thanks @pruivo. I made some other changes.
I would avoid javadoc for `*Processor` class.
Let me know if we can improve something else here.

yrodiere commented:
> Changed the FullText annotation to Text, I had a chat with @karesti yesterday, probably less is more...

Just a warning: text fields are not "just" text, they really are full-text, which implies features (text analysis, tokenization, ...) but also limitations (no sorting, no aggregations, ...). In that way, just naming the annotation "Text" may lead to confusion.

That's all, I warned y'all, I'll be on my way :)

yrodiere commented on api/src/main/java/org/infinispan/api/annotations/indexing/Id.java:
```diff
@@ -0,0 +1,27 @@
+@Target({ElementType.METHOD, ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+@PropertyMapping(processor = @PropertyMappingAnnotationProcessorRef(type = IdProcessor.class, retrieval = BeanRetrieval.CONSTRUCTOR))
+public @interface Id {
```
Are you sure this is necessary in Infinispan? I thought the ID was always the cache key?

fax4ever commented:
Actually the Lucene document ID is uniquely derived in both senses from the corresponding cache entry key. The algorithm is written in the `KeyTransformationHandler` class that is used by the `KeyTransformationHandlerIdentifierBridge`. 
In that sense we cannot make the user choose their own id, I think. I added `@Id` annotation since I remember that Adrian was working on queries by id, but let's not add from now on. I'm removing the annotation, thanks @yrodiere.

About the `@Text` annotation. Probably I’m obsessed with making things as minimal as possible. I think that word `text` is clear enough and is word very used in the context of the indexed queries. I think that also ES uses `text` to define this field type. 
Anyway it would be OK for me if we decided to use another word.

tristantarrant commented:
I would suggest adding some "demo" entities annotated with all of the possible annotations.

pruivo reviewed (approved):
LGTM

fax4ever commented:
> I would suggest adding some "demo" entities annotated with all of the possible annotations.

good idea! done with the last push

tristantarrant commented on api/src/main/java/org/infinispan/api/common/annotations/indexing/BasicProcessor.java:
```diff
@@ -0,0 +1,29 @@
+public class BasicProcessor implements PropertyMappingAnnotationProcessor<Basic> {
+
+   @Override
+   public void process(PropertyMappingStep mapping, Basic annotation, PropertyMappingAnnotationProcessorContext context) {
```
Could we move these `processors` to another module ? I don't like the fact that such code is delivered as part of an API module.

fax4ever replied:
yes the idea is to have these classes in a query module for instance

tristantarrant commented:
LGTM. Just remove the processors and we can merge

fax4ever commented:
@tristantarrant moved the processors to a different package, of course you can squash my commits before the commit

tristantarrant commented:
Merged, thanks

fax4ever commented:
Thanks!
